### PR TITLE
Properly utilize CircleCI caches per build

### PR DESCRIPTION
**Pull request by danpat**
# Issue

CircleCI cache keys are immutable.  Once you've written to a cache key, it can't be changed until it expires (30 days).

We are currently trying to "save the cache" to a key called `ccache` for all builds.  This doesn't really work, as most builds fail to write to the cache, and the restored cache that most builds get doesn't contain any data that leads to `ccache` hits.

The fix is to use [partial cache restore](https://circleci.com/docs/2.0/caching/#restoring-cache).  The basic strategy is:

  1. When a build completes, write out to a very specific cache key.  In our case, the build name, plus the branch, and the current timestamp.
  2. When a build tries to restore the cache, try a series of fallbacks:
    1. Use the build name+branch+timestamp, or
    2. Use the most recent build name+branch-* key
    3. Use the most recent build name-* key

This should have the effect of actually getting some cache hits - it'll pick up the most recent completed build on a branch if available, otherwise it'll pick up the most recent build of that type (e.g. `ccache-release-linux-x86_64`).  This can mean that the `.ccache` that gets restored isn't super valid, but it should be better than nothing for most builds.  For branches that only get one commit, the hit rate may be low, but we should see improved hit rates for longer-lived branches where many commits are made and the `.ccache` from a previous build contains useful cache results.

Overall, the hope is that this should bring down the build times on CI for many common cases.

One of the focuses here is to improve the build times for tag builds - often when those occur, we want them to be done as quickly as possible (as a previous "testing" build has already completed, so it's mostly just a matter of re-building with the tag).

`build-release` goes from about 10+ minutes with no cache hits down to about 4m30s for a no-op build (full cache hits).
`lint-linux-debug` (the slowest job) goes from 29 minutes down to about 21 minutes.

This change only speeds up compilation steps using `ccache`.  Further work could be done to detect when binaries are unchanged and tests do not need to be re-run (for example).

## Tasklist

 - [ ] Update the [changelog](CHANGELOG.md)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -70,13 +63,18 @@ jobs:
       - run: git submodule sync && git submodule update --init
       - restore_cache:
           keys:
-            - ccache
+            - ccache-release-linux-x86_64-{{ .Branch }}
```
Comment (PureW) [inline .circleci/config.yml]: Does circleci grab the first cache that exists?

File: .circleci/config.yml [hunk 2]
```diff
@@ -150,17 +146,19 @@ jobs:
             && cmake .. -DCMAKE_BUILD_TYPE=Debug -DENABLE_COVERAGE=On -DCPACK_GENERATOR=DEB \
                 -DENABLE_COMPILER_WARNINGS=On -DENABLE_WERROR=Off -DCMAKE_EXPORT_COMPILE_COMMANDS=On
       - run: make -C build -j4
-      - run: scripts/clang-tidy-only-diff.sh 4
       - run: make -C build -j2 tests
       - run: make -C build -j2 check
+      # Note: we save the cache here before doing linting so that if linting fails, we can rebuild quickly
```
Comment (PureW) [inline .circleci/config.yml]: Should we have a separate build target for linting? Many times linting is the culprit and having to wait for a build, even if partially cached, would waste some time.

**Comment by PureW** [comment]
Just as a potential alternative, I've used `sccache` in the past which is `ccache` but support aws s3 backend. https://github.com/mozilla/sccache/

File: .circleci/config.yml (hunk 1 above)
Comment (PureW) [reply .circleci/config.yml]: > If there are multiple matches, the most recently generated cache will be used.

yep.

File: .circleci/config.yml [hunk 3]
```diff
@@ -180,12 +179,12 @@ jobs:
       - run: make -C build -j4
       - run: make -C build -j4 tests
       - run: make -C build -j2 check
-      - run: make -C build install
-      - run: make -C build package
       - save_cache:
-          key: ccache
+          key: ccache-release-linux-x86_64-{{ .Branch }}-{{ epoch }}
```
Comment (PureW) [inline .circleci/config.yml]: Is `epoch` current time? Isn't this always unique? When would this actually be a cache hit?

File: .circleci/config.yml (hunk 1 above)
Comment (danpat) [reply .circleci/config.yml]: Yeah - it's a series of fallbacks - this should cause it to grab the most recent cache that matches this particular pattern.

File: .circleci/config.yml (hunk 2 above)
Comment (danpat) [reply .circleci/config.yml]: Yeah, it's the slowest part - I need to do more thinking about how we can refactor the CI workflow to not get blocked by that.

File: .circleci/config.yml (hunk 3 above)
Comment (danpat) [reply .circleci/config.yml]: Yes - this is the only way to make sure we write out a new cache dump with each build.  Once a cache key is created, it's immutable for 30 days, so each new build needs to write to a new cache.

The real trick here is the `read_cache` rules, which will find the *most recent* cache key that matches the desired pattern.  That's how a cache from the previous build will get used in the next one.

After reading the CircleCI docs a few times, it seems that this is how its intended to work, but it's not _quite_ spelled out clearly, just strongly implied.

**Comment by danpat** [comment]
@PureW yeah, the only problem with `sccache` is that we need a spot to actually store the shared cache itself - we'd have to find a willing donor to supply an S3 bucket we can use.

**Pull request merged**